rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,588 @@
+  - Need to add the command to the reference guide.
+- List of limitation of the up-to-date implementation.
+
+## Drawbacks
+
+- Editing the custom fields is not possible, since it would require data migration. However, this is the same case as for any existing field.
+
+## Alternatives
+
+### Database vs Static file configuration
+
+The different options between having a runtime configuration stored in the DB and defining everything as a static python configuration were analyzed and below the resuls are presented.
+
+**DB configuration**
+
+Storing the configuration options in the database would enable instance administrators to dynamically select and add new custom fields available globally in the system. That said, adding a new custom field can have a big impact on the system in terms of e.g how to index, serialize in different formats etc. Moreover, the use case of configuring the system's custom fields is considered a task that most probably is executed during the first steps of an instance's deployment meaning that the impact from adding them after the first launch is not high.
+
+**Static file configuration**
+
+The preferred way to store the configuration regarding custom fields is in a python configuration file. With that approach the configuration is:
+
+- Consistent along invenio ecosystem, following the same patterns all the other configuration is loaded when the application is loaded.
+- Provides flexibility in what a system admin can configure e.g use programmatic API.
+
+**Conclusion**
+
+Configuration will be done in cfg/python files as is the pattern in Invenio. However, the advantages of DB configuration have been noted and a wider task to implement it might be carried out in the future.
+
+### Nested vs Object as Elasticsearch mapping
+
+The question is about storing the custom fields as `object` or `nested`. The difference between them is the search behaviour and performance on objects with lists. Imagine there is the need to add an object `measures` which contain `width` and `heigth`. In the mapping it is represented as follows:
+
+```json
+{
+  "measures": [
+    {
+      "heigth": 100,
+      "width": 100
+    },
+    {
+      "heigth": 50,
+      "width": 50
+    }
+  ]
+}
+```
+
+Using `nested` field each measure is stored separately. However, `object` would flatten the values to:
+
+```
+{
+    "measures.heigth": [100, 50],
+    "measures.width": [100, 50]
+}
+```
+
+As a consequence, querying for "height=100 AND width=50" would produce a match, when in reality there is no measure that matches it.
+
+**Consequences**
+
+While using `nested` fields improves the quality of search, it can have a big impact in performance. Some cases where queries slow down from 5ms to 240ms [have been reported](https://discuss.elastic.co/t/performance-on-query-with-nested-objects/181792). Moreover, the current query parser does not support nested queries. While code can be ported from Zenodo, it will have added costs in people resources.
```

Comment:
Question: Do custom fields on Zenodo use nested? If so, do we know if there's a requirement from Plazi to do specific searches?

Reply:
In Zenodo they are nested, but during the initial discussions, it was deemed unnecessary.

After discussions with @slint: There is no requirement from Plazi.